---
title: Cuffdiff Analysis
output: html_document
params:
   cuffdiff_path: ''
   save_plots: TRUE
   save_plots_path: ''
---

```{r setup}
library(seqc)
library(magrittr)
library(ggplot2)
# Debug outputs
params$cuffdiff_path %>% print()
params$save_plots %>% print()
params$save_plots_path %>% print()
# End debug outputs
saveAndPrintWrap <- function(plot, plot_name) {
  plot %>% seqc::saveAndPrint(save_plot = params$save_plots, plot_name = plot_name, output_path = params$save_plots_path)
}
cuff <- read_cuffdiff(params$cuffdiff_path)
```

This Notebook is guided by the [cummeRbund Manual](https://www.bioconductor.org/packages/3.7/bioc/vignettes/cummeRbund/inst/doc/cummeRbund-manual.pdf)

# Global Statistics & Quality Control

```{r load_data}
cuff %>% print()
```

> Several plotting methods are available that allow for quality-control or global analysis of cufflinks data. A good place to begin is to evaluate the quality of the model fitting. Overdispersion is a common problem in RNA-Seq data. As of cufflinks v2.0 mean counts, variance, and dispersion are all emitted, allowing you to visualize the estimated overdispersion for each sample as a quality control measure.

```{r dispersion_plot}
cuff %>% genes() %>% dispersionPlot() %>%
  saveAndPrintWrap('dispersionPlot')
```

> The squared coefficient of variation is a normalized measure of cross-replicate variability that can be useful for evaluating the quality your RNA-seq data. Differences in CV 2 can result in lower numbers of differentially expressed genes due to a higher degree of variability between replicate fpkm estimates.

```{r fpkmSCVPlot}
cuff %>% genes() %>% fpkmSCVPlot() %>% 
  saveAndPrintWrap('genes_scv')
cuff %>% isoforms() %>% fpkmSCVPlot() %>%
  saveAndPrintWrap('isoforms_scv')
```

> To assess the distributions of FPKM scores across samples, you can use the csDensity plot

```{r csdensity}
cuff %>% genes() %>% csDensity() %>%
  saveAndPrintWrap('csDensity')
cuff %>% genes() %>% csDensity(replicates = TRUE) %>%
  saveAndPrintWrap('csDensityReplicates')
```